### PR TITLE
LibGL: Extract software renderer from LibGL into separate library LibSoftGPU

### PR DESCRIPTION
**Pull request by sunverwerth**
The goal of this PR is to extract everything related to software rendering from LibGL and put it into its own library LibSoftGPU.
Afterwards we can start working on a generic user mode driver interface so that LibGL can eventually make use of hardware acceleration once drivers for that land in the project.

### CODE REVIEW

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 1]
```diff
@@ -208,175 +207,53 @@ void SoftwareGLContext::gl_end()
...
-            tri.vertices[2] = m_clipped_vertices[i + 1];
-            processed_triangles.append(tri);
-        }
-    }
-
-    m_bound_texture_units.clear();
-    for (auto& texture_unit : m_texture_units) {
-        if (texture_unit.is_bound())
-            m_bound_texture_units.append(texture_unit);
+    Vector<size_t, 32> enabled_texture_units;
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:225]: Can we centralize the total number of texture units (32) in a `static constexpr` somewhere?

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 2]
```diff
@@ -2596,4 +2635,107 @@ void SoftwareGLContext::present()
 {
     m_rasterizer.blit_to(*m_frontbuffer);
 }
+
+void SoftwareGLContext::sync_device_config()
+{
+    if (m_sampler_config_is_dirty) {
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:2641]: This `if` should be inverted (early return) to reduce nesting

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 3]
```diff
@@ -2596,4 +2635,107 @@ void SoftwareGLContext::present()
...
+        for (int i = 0; i < 32; ++i) {
+            SoftGPU::SamplerConfig config;
+            auto texture = m_texture_units[i].bound_texture_2d();
+            if (m_texture_units[i].texture_2d_enabled() && texture) {
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:2645]: This `if` could be inverted as well, to remove another nesting level

File: Userland/Libraries/LibGL/SoftwareGLContext.h [hunk 4]
```diff
@@ -176,10 +179,7 @@ class SoftwareGLContext : public GLContext {
     FloatVector4 m_current_vertex_tex_coord = { 0.0f, 0.0f, 0.0f, 1.0f };
     FloatVector3 m_current_vertex_normal = { 0.0f, 0.0f, 1.0f };
 
-    Vector<GLVertex, 96> vertex_list;
-    Vector<GLTriangle, 32> triangle_list;
-    Vector<GLTriangle, 32> processed_triangles;
-    Vector<GLVertex> m_clipped_vertices;
+    Vector<SoftGPU::Vertex, 96> vertex_list;
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.h:182]: This should probably be `m_vertex_list` (old code had it wrong as well)

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 5]
```diff
@@ -487,41 +491,188 @@ static Gfx::IntSize closest_multiple(const Gfx::IntSize& min_size, size_t step)
...
+
+            // to screen space
+            vec.position.set_x(scr_width / 2 + vec.position.x() * scr_width / 2);
+            vec.position.set_y(scr_height / 2 - vec.position.y() * scr_height / 2);
+        }
+
+        Triangle tri;
+        tri.vertices[0] = m_clipped_vertices[0];
+        for (size_t i = 1; i < m_clipped_vertices.size() - 1; i++) {
+
```
Comment (GMTA) [inline Userland/Libraries/LibSoftGPU/Device.cpp:616]: Superfluous newline

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 6]
```diff
@@ -647,4 +798,22 @@ float SoftwareRasterizer::get_depthbuffer_value(int x, int y)
...
+    VERIFY(depth > 0);
+    VERIFY(levels > 0);
+    VERIFY(layers > 0);
+
+    return adopt_ref(*new Image(format, width, height, depth, levels, layers));
+}
+
+void Device::set_sampler_config(unsigned sampler, SamplerConfig const& config)
+{
+    VERIFY(sampler < num_samplers);
```
Comment (GMTA) [inline Userland/Libraries/LibSoftGPU/Device.cpp:814]: If you make `m_samplers` an `AK::Array`, you get these checks automatically

**Review by GMTA — COMMENTED** [review]
Some initial findings!

**Comment by GMTA** [comment]
Testing this branch shows no regressions in `3DFileViewer`, `glquake` or `scummvm` and it fixes the jagged edges issues in the ScummVM main menu which is :yakkiss:

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp (hunk 1 above)
Comment (sunverwerth) [reply Userland/Libraries/LibGL/SoftwareGLContext.cpp:225]: Agreed!

File: Userland/Libraries/LibGL/SoftwareGLContext.h (hunk 4 above)
Comment (sunverwerth) [reply Userland/Libraries/LibGL/SoftwareGLContext.h]: done

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 7]
```diff
@@ -208,175 +210,53 @@ void SoftwareGLContext::gl_end()
...
-        }
+    m_vertex_list.clear_with_capacity();
 
-        if (area > 0) {
-            swap(triangle.vertices[0], triangle.vertices[1]);
-        }
-
-        m_rasterizer.submit_triangle(triangle, m_bound_texture_units);
-    }
+    m_in_draw_state = false;
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:259]: This reintroduces the bug fixed by 12c4491cca7909daee9fbd5be822eee8d8757d91

File: Userland/Libraries/LibGL/SoftwareGLContext.cpp [hunk 8]
```diff
@@ -2596,4 +2638,115 @@ void SoftwareGLContext::present()
...
+            config.fixed_function_texture_env_mode = SoftGPU::TextureEnvMode::Modulate;
+            break;
+        case GL_REPLACE:
+            config.fixed_function_texture_env_mode = SoftGPU::TextureEnvMode::Replace;
+            break;
+        case GL_DECAL:
+            config.fixed_function_texture_env_mode = SoftGPU::TextureEnvMode::Decal;
+            break;
+        default:
+            VERIFY_NOT_REACHED();
```
Comment (GMTA) [inline Userland/Libraries/LibGL/SoftwareGLContext.cpp:2745]: To be consistent, the other `switch` blocks above should also get a default section - it looks weird to have it only here.

File: Userland/Libraries/LibGL/SoftwareGLContext.h [hunk 9]
```diff
@@ -176,10 +180,7 @@ class SoftwareGLContext : public GLContext {
     FloatVector4 m_current_vertex_tex_coord = { 0.0f, 0.0f, 0.0f, 1.0f };
     FloatVector3 m_current_vertex_normal = { 0.0f, 0.0f, 1.0f };
 
-    Vector<GLVertex, 96> vertex_list;
-    Vector<GLTriangle, 32> triangle_list;
-    Vector<GLTriangle, 32> processed_triangles;
-    Vector<GLVertex> m_clipped_vertices;
+    Vector<SoftGPU::Vertex, 96> m_vertex_list;
```
Comment (Quaker762) [inline Userland/Libraries/LibGL/SoftwareGLContext.h:183]: This 96 feels really arbitrary. I can't quite remember why it's exactly this size. Should we perhaps put it into a `constexpr` if it's related to the standard?

File: Userland/Libraries/LibSoftGPU/DeviceInfo.h [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace SoftGPU {
+
+struct DeviceInfo final {
+    String vendor_name;
+    String device_name;
+    unsigned num_texture_units;
+};
+
+}
```
Comment (Quaker762) [inline Userland/Libraries/LibSoftGPU/DeviceInfo.h:19]: Missing a newline at EOF here

File: Userland/Libraries/LibSoftGPU/Enums.h [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+enum class PrimitiveType {
+    Triangles,
+    TriangleStrip,
+    TriangleFan,
+    Quads,
+};
+
+}
```
Comment (Quaker762) [inline Userland/Libraries/LibSoftGPU/Enums.h:71]: And here

**Review by Quaker762 — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGL/SoftwareGLContext.h (hunk 9 above)
Comment (sunverwerth) [reply Userland/Libraries/LibGL/SoftwareGLContext.h:183]: I will remove the 96 since we always `clear_with_capacity()` this vector

**Review by Quaker762 — APPROVED** [review]

**Comment by Quaker762** [comment]
Looks good! L :clap: G :clap: T :clap: M

**Pull request merged**